### PR TITLE
Add set_button_text() to create text buttons

### PR DESCRIPTION
**Pull request by markmandel**
Adds the functionality to be able to dynamically generate button images and apply them based on text with a given font.

This includes text and background colouring, text scaling and positioning and support for line height over line breaks.

Adds dependencies on:
imageproc = "0.20.0"
rusttype = "0.8.3"

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -272,6 +275,33 @@ impl StreamDeck {
...
+        x: u32,
+        y: u32,
+        scale: Scale,
+        line_height: f32,
+        font: &Font,
+        text: &str,
+    ) -> Result<(), Error> {
+        let (width, height) = self.kind.image_size();
+        let background = Rgb([background.r, background.g, background.b]);
+        let colour = Rgb([colour.r, colour.g, colour.b]);
```
Comment (markmandel) [inline src/lib.rs:294]: I contemplated adding a `to_rgb()` and maybe a `to_rgba(opacity)` to Colour to clean up some of this repetition, but figured that might be best handled by a different PR.

WDYT?

**Comment by markmandel** [comment]
image

Picture of it running white text of a grey background, with a `\n` linebreak with 1.1 line height.

(And it's sitting next to my hangouts button :smile: )

File: src/lib.rs (hunk 1 above)
Comment (ryankurte) [reply src/lib.rs:294]: seems like a good improvement, up to you whether you included it here or in another pr, i don't particularly mind either way ^_^

Comment (markmandel) [reply src/lib.rs:294]: I'll do it in another PR. I'm sure I'll come back around to this at some point int the future :smile: no need to hold things up.

File: src/lib.rs [hunk 2]
```diff
@@ -272,6 +275,33 @@ impl StreamDeck {
...
+    /// Sets a button to the provided text.
+    /// Will break text over \n linebreaks
+    pub fn set_button_text(
+        &mut self,
+        key: u8,
+        colour: &Colour,
+        background: &Colour,
+        x: u32,
+        y: u32,
+        scale: Scale,
```
Comment (ryankurte) [inline src/lib.rs:287]: would it be reasonable to pack the less-used options here into a `TextOptions` object with a `Default` impl? i'm not sure where exactly the line is, but, anything shared between calls to set sequential buttons (or the same button in different states) might be a useful guideline.

not intended to be prescriptive in any way, but something along the lines of `set_button_test(&mut self, text: &str, fg: &Colour, bg: &Colour, options: &TextOptions)` that elides any customisation that isn't _strictly_ required through `TextOptions::default()`.

File: src/lib.rs [hunk 3]
```diff
@@ -272,6 +275,33 @@ impl StreamDeck {
...
+        font: &Font,
+        text: &str,
+    ) -> Result<(), Error> {
+        let (width, height) = self.kind.image_size();
+        let background = Rgb([background.r, background.g, background.b]);
+        let colour = Rgb([colour.r, colour.g, colour.b]);
+        let mut image = ImageBuffer::from_pixel(width as u32, height as u32, background);
+        let mut y = y;
+        text.to_string().split("\n").for_each(|txt| {
+            draw_text_mut(&mut image, colour, x, y, scale, font, txt);
```
Comment (ryankurte) [inline src/lib.rs:298]: i wonder if is it possible to automatically centre (or more generally align, though i think centring would be enough) this in a useful way so that people don't have to deal with the x/y offsets for a given piece of text?

**Comment by ryankurte** [comment]
oh hey this looks super neat, thanks for the pr! i left a couple of comments / questions / ideas.

(one day i'll get to adding a radius option to everything to clean up the corners for non-transparent images/backgrounds 😅)

File: src/lib.rs (hunk 3 above)
Comment (markmandel) [reply src/lib.rs:298]: I can't work out how to find out the width of the text :cry:

File: src/lib.rs (hunk 2 above)
Comment (markmandel) [reply src/lib.rs:287]: Ooh, that's an interesting idea. Let me have think, and will update the PR.

File: src/lib.rs (hunk 3 above)
Comment (ryankurte) [reply src/lib.rs:298]: Hmm, doesn't seem super easy! I _think_ it should be possible using something like [`Font::layout`](https://docs.rs/rusttype/0.9.2/rusttype/enum.Font.html#method.layout) or the via looking at the glyphs manually.

To avoid this for now we can put a `Position` enum in the `ImageOptions` object with an `Absolute{x: .., y: ..}` variant so adding this later just involves an additional variant and is thus a non-breaking change.

Comment (markmandel) [reply src/lib.rs:298]: That's a fantastic idea. :+1:

File: src/lib.rs [hunk 4]
```diff
@@ -272,6 +276,34 @@ impl StreamDeck {
...
 
+    /// Sets a button to the provided text.
+    /// Will break text over \n linebreaks
+    pub fn set_button_text(
+        &mut self,
+        key: u8,
+        font: &Font,
+        pos: &TextPosition,
+        text: &str,
+        opts: &TextOptions,
```
Comment (markmandel) [inline src/lib.rs:287]: This is where I ended up with for defaults:
* I wasn't sure if we wanted to ship a font (Roboto?) with this library, so I left it as a required field. I figure users will have SOME font they like on their machine.
* I created a default size, and line height I figured look decent
* White text on a black background is always a good starting point I figured.

Rearranged the ordering a little too.

WDYT?

Comment (ryankurte) [reply src/lib.rs:287]: seems good to me! good point on the font, defaults sound good, and if positioning wasn't critical i'd probably put it in the TextOptions but since it is, might as well stay out.

it's not critical / required here but it would be excellent to have an example using this (pulling a font from somewhere? idk) if you were interested in another PR.
